lib/oli_web/live/sections/overview_view.ex

```diff
@@ -144,4 +150,65 @@ defmodule OliWeb.Sections.OverviewView do
 
     {:noreply, push_redirect(socket, to: Routes.delivery_path(socket, :index))}
   end
+
+  def handle_event("show_delete_modal", _params, socket) do
+    section_has_student_enrollments =
+      Sections.list_student_enrollments(socket.assigns.section.slug) != []
+
+    {message, action} =
+      if section_has_student_enrollments do
+        {"""
+           This section has student data and will be archived rather than deleted.
+           Are you sure you want to archive it? You will no longer have access to the data.
+         """, "Archive"}
+      else
+        {"""
+           This action cannot be undone. Are you sure you want to delete this section?
+         """, "Delete"}
+      end
+
+    modal = %{
+      component: DeleteModalNoConfirmation,
+      assigns: %{
+        id: "delete_section_modal",
+        description: message,
+        entity_type: "section",
+        entity_id: socket.assigns.section.id,
+        delete_enabled: true,
+        delete: "delete_section",
+        modal_action: action
+      }
+    }
+
+    {:noreply,
+     assign(socket, modal: modal, section_has_student_enrollments: section_has_student_enrollments)}
```

Comment:
`section_has_student_enrollments` should not be cached as an assign here, and then later used in the `delete_section` event.   This introduces a timing bug. Imagine the following:

1. Admin user clicks "Delete section"
2. The `show_delete_modal` event is handled and the check to see if section has student enrollments executes.  It returns `false`, indicating that a hard delete is possible.  The modal displays and the `section_has_student_enrollments` value is cached in the assign.
3. While the modal is displayed and before the user clicks "Ok" to proceed, a student launches from their LMS and enrolls in the section and visits the first page, thus generating student data.
4. The admin user clicks "Ok" to hard delete the section, which results in student data loss.

Reply:
This is an edge case but it might happen, good catch! If we don't cache `section_has_student_enrollments` in the assigns, then we'll need to call the function again in the `delete_section` event, but in the scenario you described we would have told the user that the section was going to be "deleted", and instead it will be "archived" which is not ideal.
One option would be to still cache it in the assigns, and in the `delete_section` event compare both boolean values. If they are different then we could re-render the page asking the user to try again or so. Thoughts?

Reply:
I think re-rendering with the update that it can now only be archived makes sense.